### PR TITLE
[Example][Refactor] Refactor GIN example

### PR DESCRIPTION
**Pull request by chang-l**
## Description
Referring to: https://github.com/dmlc/dgl/pull/4186, this PR is for refactoring GIN example. Only single GPU is implemented.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [X] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [X] Changes are complete (i.e. I finished coding on this PR)
- [X] All changes have test coverage
- [X] Code is well-documented
- [X] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

## Changes
- Removed all cmd arguments, except `--dataset`
- The implementation is focused/hard-coded on: (1) eps=0 (not learning eps); (2) both neighbor and graph are enforced to do sum-pooling; (3) 5layer GCN + 2layer MLP; (4) 10-fold split of dataset; (5) report both train and validation accuracy due to small dataset.
- Removed `ApplyNodeFunc` class, use mlp instead, to feed into GINConv, referring to this issue https://github.com/dmlc/dgl/issues/4281
- Updated Readme file with test results

### CODE REVIEW

File: examples/pytorch/gin/README.md [hunk 1]
```diff
@@ -4,80 +4,23 @@ Graph Isomorphism Network (GIN)
...
 
-Dependencies
-------------
-- PyTorch 1.1.0+
-- sklearn
-- tqdm
-
-``bash
-pip install torch sklearn tqdm
-``
```
Comment (mufeili) [inline examples/pytorch/gin/README.md:15]: I think `sklearn` is still a dependency.

File: examples/pytorch/gin/README.md [hunk 2]
```diff
@@ -4,80 +4,23 @@ Graph Isomorphism Network (GIN)
...
-python main.py --dataset REDDITBINARY --device 0  \
-                --graph_pooling_type mean --neighbor_pooling_type sum --filename REDDITBINARY.txt --fold_idx 6 --epoch 120
-
-python main.py --dataset REDDITMULTI5K --device 0  \
-                --graph_pooling_type mean --neighbor_pooling_type sum --filename REDDITMULTI5K.txt
+python3 train.py --dataset MUTAG
 ```
 
-one fold of 10 result are below.
+> **_NOTE:_**  Users may observe results fluctuate due to the randomness of relatively small dataset.  In consistence with the original [paper ](https://arxiv.org/abs/1810.00826), five social network datasets, 'COLLAB', 'IMDBBINARY' 'IMDBMULTI' 'REDDITBINARY' and 'REDDITMULTI5K', are also available as the input. Users are encouraged to update the script slightly for social network applications, for example, replacing sum readout on bioinformatics datasets to mean readout on social datasets.
```
Comment (mufeili) [inline examples/pytorch/gin/README.md:16]: "randomness of relatively small dataset" -> "randomness with relatively small datasets"

Comment (mufeili) [inline examples/pytorch/gin/README.md:16]: extra empty space after "[paper"

Comment (mufeili) [inline examples/pytorch/gin/README.md:16]: extra empty space after "[paper"

Comment (mufeili) [inline examples/pytorch/gin/README.md:16]: Perhaps you also want to mention "--degree_as_nlabel"

File: examples/pytorch/gin/README.md [hunk 3]
```diff
@@ -4,80 +4,23 @@ Graph Isomorphism Network (GIN)
...
-| REDDITMULTI5K | 54.8      | 57.5 ± 1.5   |
 
+Summary
+-------
+| Dataset       | Result 
+| ------------- | -------
+| MUTAG         | ~89.4    
+| PTC           | ~68.5   
+| NCI1          | ~82.9    
+| PROTEINS      | ~74.1
```
Comment (mufeili) [inline examples/pytorch/gin/README.md:26]: Is this averaged based on k-fold?

File: examples/pytorch/gin/train.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--dataset', type=str, default="MUTAG",
+                        choices=['MUTAG', 'PTC', 'NCI1', 'PROTEINS'],
+                        help='name of dataset (default: MUTAG)')
+    args = parser.parse_args()
+    print(f'Training with DGL built-in GINConv module with a fixed epsilon = 0')
+    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+    
+    # load and split dataset
+    dataset = GINDataset(args.dataset, True, False) #  self_loop and disable one-hot encoding for input features
```
Comment (mufeili) [inline examples/pytorch/gin/train.py:126]:
```suggestion
    dataset = GINDataset(args.dataset, self_loop=True, degree_as_nlabel=False) #  self_loop and disable one-hot encoding for input features
```

File: examples/pytorch/gin/train.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+            hidden_rep.append(h)
+        score_over_layer = 0
+        # perform graph sum pooling over all nodes in each layer
+        for i, h in enumerate(hidden_rep):
+            pooled_h = self.pool(g, h)
+            score_over_layer += self.drop(self.linear_prediction[i](pooled_h))
+        return score_over_layer
+    
+def split_fold10(labels, fold_idx=0, seed=0, shuffle=True):
+    skf = StratifiedKFold(n_splits=10, shuffle=shuffle, random_state=seed)
```
Comment (mufeili) [inline examples/pytorch/gin/train.py:68]: Perhaps you don't need to have `seed=0, shuffle=True` in the arguments of `split_fold10`.

File: examples/pytorch/gin/train.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+    
+    # load and split dataset
+    dataset = GINDataset(args.dataset, True, False) #  self_loop and disable one-hot encoding for input features
+    labels = [l for _, l in dataset]
+    train_idx, val_idx = split_fold10(labels)
+    
+    # create dataloader
+    train_loader = GraphDataLoader(dataset, sampler=SubsetRandomSampler(train_idx),
+                                   batch_size=128, pin_memory=torch.cuda.is_available())
```
Comment (mufeili) [inline examples/pytorch/gin/train.py:132]: I think you need to specify `shuffle=True` for `train_loader`.

File: examples/pytorch/gin/train.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    val_loader = GraphDataLoader(dataset, sampler=SubsetRandomSampler(val_idx),
+                                 batch_size=128, pin_memory=torch.cuda.is_available(), shuffle=False)
+    
+    # create GIN model
+    in_size = dataset.dim_nfeats
+    out_size = dataset.gclasses
+    model = GIN(in_size, 16, out_size).to(device)
```
Comment (mufeili) [inline examples/pytorch/gin/train.py:139]: I suppose `hidden_dim` used to be `64` from `ginparser.py`?

File: examples/pytorch/gin/train.py [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+        self.batch_norm = nn.BatchNorm1d((hidden_dim))
+
+    def forward(self, x):
+        h = x
+        h = F.relu(self.batch_norm(self.linears[0](h)))
+        return self.linears[1](h)
+    
+class GIN(nn.Module):
+    def __init__(self, input_dim, hidden_dim, output_dim):
+        super().__init__()
```
Comment (mufeili) [inline examples/pytorch/gin/train.py:31]: You can first specify `self.num_layers=5`, which makes the code more readable.

File: examples/pytorch/gin/train.py [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+        self.ginlayers = torch.nn.ModuleList()
+        self.batch_norms = torch.nn.ModuleList()
```
Comment (mufeili) [inline examples/pytorch/gin/train.py:33]: "torch.nn" -> "nn" for both L32 and 33

File: examples/pytorch/gin/train.py [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+from dgl.dataloading import GraphDataLoader
+from dgl.nn.pytorch.conv import GINConv
+from dgl.nn.pytorch.glob import SumPooling
+import argparse
+
+class MLP(nn.Module):
+    """Construct two-layer MLP-type aggreator for GIN model"""
+    def __init__(self, input_dim, hidden_dim, output_dim):
+        super().__init__()
+        self.linears = torch.nn.ModuleList()
```
Comment (mufeili) [inline examples/pytorch/gin/train.py:18]: Ditto for `torch.nn`

File: examples/pytorch/gin/train.py [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+        super().__init__()
+        self.ginlayers = torch.nn.ModuleList()
+        self.batch_norms = torch.nn.ModuleList()
+        # five-layer GCN with two-layer MLP aggregator and sum-neighbor-pooling scheme
+        for layer in range(5 - 1): # excluding the input layer
+            if layer == 0:
+                mlp = MLP(input_dim, hidden_dim, hidden_dim)
+            else:
+                mlp = MLP(hidden_dim, hidden_dim, hidden_dim)
+            self.ginlayers.append(GINConv(mlp, learn_eps=False)) # set to True if learning epsilon
```
Comment (mufeili) [inline examples/pytorch/gin/train.py:40]: Why did you only include `learn_eps` given that its default value is already False?

File: examples/pytorch/gin/train.py [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+            self.batch_norms.append(nn.BatchNorm1d(hidden_dim))
+        # linear functions for graph sum poolings of output of each layer
+        self.linear_prediction = torch.nn.ModuleList()
```
Comment (mufeili) [inline examples/pytorch/gin/train.py:43]: Ditto for `torch.nn`

File: examples/pytorch/gin/train.py [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+        for layer in range(5):
+            if layer == 0:
+                self.linear_prediction.append(nn.Linear(input_dim, output_dim))
+            else:
+                self.linear_prediction.append(nn.Linear(hidden_dim, output_dim))
+        self.drop = nn.Dropout(0.5)
+        self.pool = SumPooling() # change to mean readout (AvgPooling) on social datasets
```
Comment (mufeili) [inline examples/pytorch/gin/train.py:50]: "social" -> "social network"

File: examples/pytorch/gin/train.py [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+    for idx in skf.split(np.zeros(len(labels)), labels):
+        idx_list.append(idx)
+    train_idx, valid_idx = idx_list[fold_idx]
+    return train_idx, valid_idx
+
+def evaluate(dataloader, device, model):
+    model.eval()
+    total = 0
+    total_correct = 0
+    for batched_graph, labels  in dataloader:
```
Comment (mufeili) [inline examples/pytorch/gin/train.py:79]: extra empty space after "labels"

File: examples/pytorch/gin/train.py [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+        batched_graph = batched_graph.to(device)
+        labels = labels.to(device)
+        feat = batched_graph.ndata.pop('attr')
+        total += len(labels)
+        logits = model(batched_graph, feat)
+        _, predicted = torch.max(logits.data, 1)
```
Comment (mufeili) [inline examples/pytorch/gin/train.py:85]: Is it necessary to do `.data`?

**Comment by mufeili** [comment]
I've done a first pass.

File: examples/pytorch/gin/README.md (hunk 2 above)
Comment (chang-l) [reply examples/pytorch/gin/README.md:16]: Thanks.

Comment (chang-l) [reply examples/pytorch/gin/README.md:16]: Sure.

Comment (chang-l) [reply examples/pytorch/gin/README.md:16]: This example is implemented more for bio datasets, where one-hot encoding is always disabled.
Now, I mention `degree_as_nlabel` in the user note, to remind users when applying to soical datasets.

File: examples/pytorch/gin/train.py (hunk 10 above)
Comment (chang-l) [reply examples/pytorch/gin/train.py:18]: Thanks.

File: examples/pytorch/gin/train.py (hunk 8 above)
Comment (chang-l) [reply examples/pytorch/gin/train.py:31]: Good point. Thanks. I will update accordingly.

File: examples/pytorch/gin/train.py (hunk 9 above)
Comment (chang-l) [reply examples/pytorch/gin/train.py:33]: Thanks.

File: examples/pytorch/gin/train.py (hunk 11 above)
Comment (chang-l) [reply examples/pytorch/gin/train.py:40]: Because whether including `eps` as a learnable parameter is an important variant in the paper. Although they didn't show any improvements of learning `eps` over fixing `eps=0`, I think it is still worthwhile to remind users there exists such option (i.e., setting `lear_eps=True`) in reflection of the orig. paper.

File: examples/pytorch/gin/train.py (hunk 12 above)
Comment (chang-l) [reply examples/pytorch/gin/train.py:43]: Thanks.

File: examples/pytorch/gin/train.py (hunk 13 above)
Comment (chang-l) [reply examples/pytorch/gin/train.py:50]: Sure.

File: examples/pytorch/gin/train.py (hunk 5 above)
Comment (chang-l) [reply examples/pytorch/gin/train.py:68]: Right. I updated.

File: examples/pytorch/gin/train.py (hunk 14 above)
Comment (chang-l) [reply examples/pytorch/gin/train.py:79]: Thanks.

File: examples/pytorch/gin/train.py (hunk 4 above)
Comment (chang-l) [reply examples/pytorch/gin/train.py:126]: Thanks. It's better.

File: examples/pytorch/gin/train.py (hunk 6 above)
Comment (chang-l) [reply examples/pytorch/gin/train.py:132]: Sure.

File: examples/pytorch/gin/train.py (hunk 7 above)
Comment (chang-l) [reply examples/pytorch/gin/train.py:139]: You are right. In `ginparser.py`, it is 64 for `hidden_dim`.
I changed it to 16 to be consistent with the [paper](https://arxiv.org/pdf/1810.00826.pdf).  They use 16 or 32 hidden units for the bio datasets, and 64 for social datasets.

File: examples/pytorch/gin/train.py (hunk 15 above)
Comment (chang-l) [reply examples/pytorch/gin/train.py:85]: Umm... I don't think we need that. Removed.

File: examples/pytorch/gin/README.md (hunk 1 above)
Comment (chang-l) [reply examples/pytorch/gin/README.md:15]: Thanks. I updated.

File: examples/pytorch/gin/README.md [hunk 16]
```diff
@@ -6,78 +6,28 @@ Graph Isomorphism Network (GIN)
...
-| MUTAG         | 89.4      | 89.4 ± 5.6   |
-| PTC           | 68.5      | 64.6 ± 7.0   |
-| NCI1          | 78.5      | 82.7 ± 1.7   |
-| PROTEINS      | 72.3      | 76.2 ± 2.8   |
-| COLLAB        | 81.6      | 80.2 ± 1.9   |
-| IMDBBINARY    | 73.0      | 75.1 ± 5.1   |
-| IMDBMULTI     | 54.0      | 52.3 ± 2.8   |
-| REDDITBINARY  | 88.0      | 92.4 ± 2.5   |
-| REDDITMULTI5K | 54.8      | 57.5 ± 1.5   |
+> **_NOTE:_**  Users may observe results fluctuate due to the randomness with relatively small dataset.  In consistence with the original [paper](https://arxiv.org/abs/1810.00826), five social network datasets, 'COLLAB', 'IMDBBINARY' 'IMDBMULTI' 'REDDITBINARY' and 'REDDITMULTI5K', are also available as the input. Users are encouraged to update the script slightly for social network applications, for example, replacing sum readout on bioinformatics datasets to mean readout on social network datasets and using one-hot encodings of node degrees by setting "degree_as_nlabel=True".
```
Comment (mufeili) [inline examples/pytorch/gin/README.md:24]: "to" -> "with"

Comment (mufeili) [inline examples/pytorch/gin/README.md:24]: "using one-hot encodings of node degrees by setting "degree_as_nlabel=True"." -> "using one-hot encodings of node degrees by setting "degree_as_nlabel=True" in GINDataset."

File: examples/pytorch/gin/train.py (hunk 6 above)
Comment (mufeili) [reply examples/pytorch/gin/train.py:133]: I see. Thanks.

File: examples/pytorch/gin/train.py (hunk 7 above)
Comment (mufeili) [reply examples/pytorch/gin/train.py:140]: I see.

File: examples/pytorch/gin/train.py (hunk 11 above)
Comment (mufeili) [reply examples/pytorch/gin/train.py:41]: Ok. Thanks.

**Review by mufeili — APPROVED** [review]
I'm good other than some final minor comments. Thanks.

File: examples/pytorch/gin/README.md (hunk 16 above)
Comment (chang-l) [reply examples/pytorch/gin/README.md:24]: Sure.

Comment (chang-l) [reply examples/pytorch/gin/README.md:24]: Okay. I make the update.
